### PR TITLE
Lazy load commands

### PR DESCRIPTION
**Pull request by alexdenvir**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description

Adds a `LazyLoadingCommand` class based on https://github.com/xtreamwayz/expressive-console/blob/master/src/LazyLoadingCommand.php and updates the `ContainerCommandLoader` classes to use this to lazy load all commands. Tests provided

### CODE REVIEW

File: test/ContainerCommandLoaderTest.php [hunk 1]
```diff
@@ -30,7 +34,24 @@ public function testGetCommandHasName()
...
     }
+
+    public function testGetCommandReturnsLazyCommand()
+    {
+        $cwd = getcwd();
+        chdir(__DIR__ . '/TestAsset');
+        $container = ContainerResolver::resolve();
+        chdir($cwd);
+
+        $config = $container->get('ApplicationConfig');;
```
Comment (samsonasik) [inline test/ContainerCommandLoaderTest.php:48]: need clean up double `;`

**Comment by weierophinney** [comment]
@alexdenvir Please see https://travis-ci.com/github/laminas/laminas-cli/jobs/319225632 — it looks like under PHP 7.1 with latest dependencies, this fails due to a signature mismatch. Can you try and identify what's happening?

**Comment by alexdenvir** [comment]
> @alexdenvir Please see [travis-ci.com/github/laminas/laminas-cli/jobs/319225632](https://travis-ci.com/github/laminas/laminas-cli/jobs/319225632) — it looks like under PHP 7.1 with latest dependencies, this fails due to a signature mismatch. Can you try and identify what's happening?

I can certainly have a look, although this is happening on the [master branch as well](https://travis-ci.com/github/laminas/laminas-cli/jobs/316744283). If I can work it out, I'll open a separate PR with a fix

**Comment by michalbundyra** [comment]
@weierophinney @alexdenvir the issue might be due to phpunit configuration (with coverage) as all classes (files) are loaded to calculate the coverage also on files which are not loaded in tests. This is just configuration issue, present on master, as mentioned, so I wouldn't bother too much about it.

**Comment by boesing** [comment]
LGTM.
However, moving all commands to a lazy loader is always a thing. I would prefer having some kind of validation in development mode or wherever so I can at least check if my configuration works as expected.

**Comment by michalbundyra** [comment]
@boesing 
> However, moving all commands to a lazy loader is always a thing. I would prefer having some kind of validation in development mode or wherever so I can at least check if my configuration works as expected.

This is actually a very good point.
I was thinking about it, and this is just for loading list of commands. And - listing all available commands - is something people are gonna do "only" in dev mode.

I see two options:
- add configuration:
```php
'laminas-cli' => [
    'list-lazy-loading'=> true,
    'commands' => [...],
],
```
or
- parameter to laminas-cli, for example: `vendor/bin/laminas-cli --validate`
or
- our command to validate all other commands: `vendor/bin/laminas-cli cli:validate-commands`

Thoughts?

**Comment by alexdenvir** [comment]
Related to #4 - a dev mode check could be used to determine if lazy loading should be used AND if dev-only commands should be loaded?

**Comment by michalbundyra** [comment]
@boesing In the other hands validating Container configuration shouldn't be part of that library...

**Comment by boesing** [comment]
> @boesing
> 
> > However, moving all commands to a lazy loader is always a thing. I would prefer having some kind of validation in development mode or wherever so I can at least check if my configuration works as expected.
> 
> This is actually a very good point.
> I was thinking about it, and this is just for loading list of commands. And - listing all available commands - is something people are gonna do "only" in dev mode.
> 
> * parameter to laminas-cli, for example: `vendor/bin/laminas-cli --validate`

I prefer `--validate` or caching the result if `command-cache` is activated instead of using "lazy" stuff.

> @boesing In the other hands validating Container configuration shouldn't be part of that library...

Yup, thats a thing. But I am always annoyed if a bug first is discovered if a specific command is being executed. Having `$container->has()` during instantiation of `LazyCommandLoader` would probably avoid having surprising in production. But not sure if thats okayish in the "lazy" context.

**Comment by geerteltink** [comment]
Just wondering... why do you need validation? You test the commands, you test the container and it all should work?

**Comment by michalbundyra** [comment]
Thanks, @alexdenvir!

Merged now as-is. If we decide to add some validation we can easily do it later :)

**Comment by boesing** [comment]
> Just wondering... why do you need validation? You test the commands, you test the container and it all should work?

Thats true. Sometimes I tend to be too defensive (which is related to my job I guess), maybe you're right and a validation is not needed.

**Pull request merged**